bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/RuleEngine.java

```diff
@@ -0,0 +1,426 @@
+        if (mtId == null) {
+            throw new IllegalArgumentException("Invalid module type id. It must not be null!");
+        }
+        int idx = mtId.indexOf(MODULE_TYPE_SEPARATOR);
+        if (idx != -1) {
+            mtId = mtId.substring(0, idx);
+        }
+        ModuleHandlerFactory msf = moduleHandlerFactories.get(mtId);
+        if (msf == null) {
+            throw new IllegalArgumentException("Invalid module handler factory: " + mtId);
+        }
+        return msf.create(m);
+    }
+
+    protected RuleImpl removeRule(String id) {
+        return rules.remove(id);
+    }
+
+    protected RuleImpl getRule(String rUID) {
+        return rules.get(rUID);
+    }
+
+    protected void setRuleEnable(String rUID, boolean isEnabled) {
+        RuleImpl r = rules.get(rUID);
+        if (r != null) {
+            r.setEnabled(isEnabled);
+        }
+    }
+
+    protected boolean isRunning(String rUID) {
+        return runningRules.get(rUID) != null;
+    }
+
+    public Collection<Rule> getRulesByTag(String tag) {
+        Collection<Rule> result = new ArrayList<Rule>(10);
+        for (Iterator<RuleImpl> it = rules.values().iterator(); it.hasNext();) {
+            RuleImpl r = it.next();
+            if (tag != null) {
+                Set<String> tags = r.getTags();
+                if (tags != null && tags.contains(tag)) {
+                    result.add(new RuleImpl(r));
+                }
+            } else {
+                result.add(new RuleImpl(r));
+            }
+        }
+        return result;
+    }
+
+    public Collection<Rule> getRules(String filter) {
+        Collection<Rule> result = new ArrayList<Rule>(10);
+        for (Iterator<RuleImpl> it = rules.values().iterator(); it.hasNext();) {
+            RuleImpl r = it.next();
+            result.add(new RuleImpl(r));
+        }
+        return result;
+    }
+
+    protected Collection<String> getScopeIds() {
+        Set<String> result = new HashSet<String>(10);
+        for (Iterator<RuleImpl> it = rules.values().iterator(); it.hasNext();) {
+            RuleImpl r = it.next();
+            String id = r.getScopeIdentifier();
+            if (id != null) {
+                result.add(id);
+            }
+        }
+        return result;
+    }
+
+    /**
+     *
+     * @see org.osgi.util.tracker.ServiceTrackerCustomizer#addingService(org.osgi.framework.ServiceReference)
+     */
+    @Override
+    public ModuleHandlerFactory addingService(ServiceReference/* <ModuleHandlerFactory> */reference) {
+        ModuleHandlerFactory msf = (ModuleHandlerFactory) bc.getService(reference);
+        Collection<String> moduleTypes = msf.getTypes();
+        for (Iterator<String> it = moduleTypes.iterator(); it.hasNext();) {
+            String moduleTypeName = it.next();
+            if (moduleHandlerFactories.get(moduleTypeName) != null) {
+                // log.error
+            } else {
+                moduleHandlerFactories.put(moduleTypeName, msf);
```

Comment:
Here we will have to analyze all rules that are inactive due to a missing handler and see if they are now ready to be used.